Instance does not show all services

Seems that a job does not show services for some reason. After looking into it we discovered that this job was having multiple task groups. Could it be that your JSON parsing code does not work properly with multiple task groups?

https://github.com/FRosner/cluster-broccoli/blob/74e56050add6ceb760c9937145433c173a6cec1a/app/de/frosner/broccoli/services/NomadService.scala#L137

```
[ERROR] NomadService - Requesting services for service-name failed: play.api.libs.json.JsResultException: JsResultException(errors:List((,List(ValidationError(List(error.expected.jsarray),WrappedArray())))))
```

--- a/app/de/frosner/broccoli/nomad/models/Job.scala
+++ b/app/de/frosner/broccoli/nomad/models/Job.scala
@@ -1,0 +1,11 @@
+package de.frosner.broccoli.nomad.models
+
+import play.api.libs.json.{JsPath, Json, Reads, Writes}
+
+case class Job(services: Option[Seq[Service]])
+
+object Job {
+
+  implicit val jobReads: Reads[Job] = (JsPath \ "Services").read[Option[Seq[Service]]].map(Job.apply)
+
+}

--- a/app/de/frosner/broccoli/nomad/models/Job.scala
+++ b/app/de/frosner/broccoli/nomad/models/Job.scala
@@ -3,7 +3,7 @@ package de.frosner.broccoli.nomad.models
 import play.api.libs.json._
 import play.api.libs.functional.syntax._
 
-case class Job(taskGroups: Option[Seq[TaskGroup]])
+final case class Job(taskGroups: Option[Seq[TaskGroup]])
 
 object Job {
 

--- a/app/de/frosner/broccoli/nomad/models/Job.scala
+++ b/app/de/frosner/broccoli/nomad/models/Job.scala
@@ -2,10 +2,10 @@ package de.frosner.broccoli.nomad.models
 
 import play.api.libs.json.{JsPath, Json, Reads, Writes}
 
-case class Job(services: Option[Seq[Service]])
+case class Job(taskGroups: Option[Seq[TaskGroup]])
 
 object Job {
 
-  implicit val jobReads: Reads[Job] = (JsPath \ "Services").read[Option[Seq[Service]]].map(Job.apply)
+  implicit val jobReads: Reads[Job] = (JsPath \ "TaskGroups").readNullable[Seq[TaskGroup]].map(Job.apply)
 
 }

--- a/app/de/frosner/broccoli/nomad/models/Job.scala
+++ b/app/de/frosner/broccoli/nomad/models/Job.scala
@@ -1,11 +1,15 @@
 package de.frosner.broccoli.nomad.models
 
-import play.api.libs.json.{JsPath, Json, Reads, Writes}
+import play.api.libs.json._
+import play.api.libs.functional.syntax._
 
 case class Job(taskGroups: Option[Seq[TaskGroup]])
 
 object Job {
 
-  implicit val jobReads: Reads[Job] = (JsPath \ "TaskGroups").readNullable[Seq[TaskGroup]].map(Job.apply)
+  implicit val jobFormat: Format[Job] =
+    (__ \ "TaskGroups")
+      .formatNullable[Seq[TaskGroup]]
+      .inmap(taskGroups => Job(taskGroups), (job: Job) => job.taskGroups)
 
 }

--- a/app/de/frosner/broccoli/nomad/models/Service.scala
+++ b/app/de/frosner/broccoli/nomad/models/Service.scala
@@ -1,14 +1,15 @@
 package de.frosner.broccoli.nomad.models
 
 import play.api.libs.json._
-import play.api.libs.json.Reads._
 import play.api.libs.functional.syntax._
 
 case class Service(name: String)
 
 object Service {
 
-  implicit val serviceReads: Reads[Service] =
-    (JsPath \ "Name").read[String].map(Service.apply)
+  implicit val serviceFormat: Format[Service] =
+    (__ \ "Name")
+      .format[String]
+      .inmap(name => Service(name), (service: Service) => service.name)
 
 }

--- a/app/de/frosner/broccoli/nomad/models/Service.scala
+++ b/app/de/frosner/broccoli/nomad/models/Service.scala
@@ -3,7 +3,7 @@ package de.frosner.broccoli.nomad.models
 import play.api.libs.json._
 import play.api.libs.functional.syntax._
 
-case class Service(name: String)
+final case class Service(name: String)
 
 object Service {
 

--- a/app/de/frosner/broccoli/nomad/models/Service.scala
+++ b/app/de/frosner/broccoli/nomad/models/Service.scala
@@ -1,0 +1,14 @@
+package de.frosner.broccoli.nomad.models
+
+import play.api.libs.json._
+import play.api.libs.json.Reads._
+import play.api.libs.functional.syntax._
+
+case class Service(name: String)
+
+object Service {
+
+  implicit val serviceReads: Reads[Service] =
+    (JsPath \ "Name").read[String].map(Service.apply)
+
+}

--- a/app/de/frosner/broccoli/nomad/models/Task.scala
+++ b/app/de/frosner/broccoli/nomad/models/Task.scala
@@ -1,12 +1,15 @@
 package de.frosner.broccoli.nomad.models
 
-import play.api.libs.json.{JsPath, Json, Reads, Writes}
+import play.api.libs.json._
+import play.api.libs.functional.syntax._
 
 case class Task(services: Option[Seq[Service]])
 
 object Task {
 
-  implicit val taskReads: Reads[Task] =
-    (JsPath \ "Services").readNullable[Seq[Service]].map(Task.apply)
+  implicit val taskFormat: Format[Task] =
+    (__ \ "Services")
+      .formatNullable[Seq[Service]]
+      .inmap(services => Task(services), (task: Task) => task.services)
 
 }

--- a/app/de/frosner/broccoli/nomad/models/Task.scala
+++ b/app/de/frosner/broccoli/nomad/models/Task.scala
@@ -1,6 +1,12 @@
 package de.frosner.broccoli.nomad.models
 
-/**
-  * Created by frosneraz on 17.07.17.
-  */
-class Task {}
+import play.api.libs.json.{JsPath, Json, Reads, Writes}
+
+case class Task(services: Option[Seq[Service]])
+
+object Task {
+
+  implicit val taskReads: Reads[Task] =
+    (JsPath \ "Services").readNullable[Seq[Service]].map(Task.apply)
+
+}

--- a/app/de/frosner/broccoli/nomad/models/Task.scala
+++ b/app/de/frosner/broccoli/nomad/models/Task.scala
@@ -3,7 +3,7 @@ package de.frosner.broccoli.nomad.models
 import play.api.libs.json._
 import play.api.libs.functional.syntax._
 
-case class Task(services: Option[Seq[Service]])
+final case class Task(services: Option[Seq[Service]])
 
 object Task {
 

--- a/app/de/frosner/broccoli/nomad/models/Task.scala
+++ b/app/de/frosner/broccoli/nomad/models/Task.scala
@@ -1,0 +1,6 @@
+package de.frosner.broccoli.nomad.models
+
+/**
+  * Created by frosneraz on 17.07.17.
+  */
+class Task {}

--- a/app/de/frosner/broccoli/nomad/models/TaskGroup.scala
+++ b/app/de/frosner/broccoli/nomad/models/TaskGroup.scala
@@ -3,7 +3,7 @@ package de.frosner.broccoli.nomad.models
 import play.api.libs.json._
 import play.api.libs.functional.syntax._
 
-case class TaskGroup(tasks: Option[Seq[Task]])
+final case class TaskGroup(tasks: Option[Seq[Task]])
 
 object TaskGroup {
 

--- a/app/de/frosner/broccoli/nomad/models/TaskGroup.scala
+++ b/app/de/frosner/broccoli/nomad/models/TaskGroup.scala
@@ -1,6 +1,12 @@
 package de.frosner.broccoli.nomad.models
 
-/**
-  * Created by frosneraz on 17.07.17.
-  */
-class TaskGroup {}
+import play.api.libs.json.{JsPath, Json, Reads, Writes}
+
+case class TaskGroup(tasks: Option[Seq[Task]])
+
+object TaskGroup {
+
+  implicit val taskGroupReads: Reads[TaskGroup] =
+    (JsPath \ "Tasks").readNullable[Seq[Task]].map(TaskGroup.apply)
+
+}

--- a/app/de/frosner/broccoli/nomad/models/TaskGroup.scala
+++ b/app/de/frosner/broccoli/nomad/models/TaskGroup.scala
@@ -1,0 +1,6 @@
+package de.frosner.broccoli.nomad.models
+
+/**
+  * Created by frosneraz on 17.07.17.
+  */
+class TaskGroup {}

--- a/app/de/frosner/broccoli/nomad/models/TaskGroup.scala
+++ b/app/de/frosner/broccoli/nomad/models/TaskGroup.scala
@@ -1,12 +1,15 @@
 package de.frosner.broccoli.nomad.models
 
-import play.api.libs.json.{JsPath, Json, Reads, Writes}
+import play.api.libs.json._
+import play.api.libs.functional.syntax._
 
 case class TaskGroup(tasks: Option[Seq[Task]])
 
 object TaskGroup {
 
-  implicit val taskGroupReads: Reads[TaskGroup] =
-    (JsPath \ "Tasks").readNullable[Seq[Task]].map(TaskGroup.apply)
+  implicit val taskGroupFormat: Format[TaskGroup] =
+    (__ \ "Tasks")
+      .formatNullable[Seq[Task]]
+      .inmap(tasks => TaskGroup(tasks), (taskGroup: TaskGroup) => taskGroup.tasks)
 
 }

--- a/app/de/frosner/broccoli/services/NomadService.scala
+++ b/app/de/frosner/broccoli/services/NomadService.scala
@@ -8,8 +8,9 @@ import javax.xml.ws.http.HTTPException
 import de.frosner.broccoli.conf
 import de.frosner.broccoli.models.JobStatus._
 import de.frosner.broccoli.models.{Instance, JobStatus, PeriodicRun}
+import de.frosner.broccoli.nomad.models.Job
 import de.frosner.broccoli.util.Logging
-import play.api.libs.json.{JsArray, JsObject, JsString, JsValue}
+import play.api.libs.json._
 import play.api.libs.ws.WSClient
 import play.api.Configuration
 
@@ -139,26 +140,16 @@ class NomadService @Inject()(configuration: Configuration, consulService: Consul
     val jobRequest = ws.url(queryUrl)
     val jobResponse = jobRequest.get().map { response =>
       if (response.status == 200) {
-        response.json.as[JsObject]
+        response.json.as[Job]
       } else {
         throw new Exception(s"Received ${response.statusText} (${response.status})")
       }
     }
-    val eventuallyJobServiceIds = jobResponse.map { jsObject =>
-      val services = (jsObject \\ "Services")
-      val serviceObjects = services.flatMap { serviceJson =>
-        val validatedServices = serviceJson.validate[JsArray].asEither
-        validatedServices match {
-          case Left(problem) =>
-            Logger.warn(s"Error when parsing service JSON in job $id: $problem")
-            List.empty
-          case Right(array) =>
-            array.value.map(_.as[JsObject])
-        }
-      }
-      serviceObjects.flatMap { serviceJsObject =>
-        serviceJsObject.value.get("Name").map(_.as[JsString].value)
-      }
+    val eventuallyJobServiceIds = jobResponse.map { job =>
+      val taskGroups = job.taskGroups.getOrElse(Seq.empty)
+      val tasks = taskGroups.flatMap(_.tasks.getOrElse(Seq.empty))
+      val services = tasks.flatMap(_.services.getOrElse(Seq.empty))
+      services.map(_.name)
     }
     val eventuallyRequestedServices = eventuallyJobServiceIds.map { jobServiceIds =>
       Logger.debug(s"${jobRequest.uri} => ${jobServiceIds.mkString(", ")}")

--- a/app/de/frosner/broccoli/services/NomadService.scala
+++ b/app/de/frosner/broccoli/services/NomadService.scala
@@ -160,14 +160,15 @@ class NomadService @Inject()(configuration: Configuration, consulService: Consul
         serviceJsObject.value.get("Name").map(_.as[JsString].value)
       }
     }
-    eventuallyJobServiceIds.onComplete {
-      case Success(jobServiceIds) =>
+    val eventuallyRequestedServices = eventuallyJobServiceIds.map { jobServiceIds =>
         Logger.debug(s"${jobRequest.uri} => ${jobServiceIds.mkString(", ")}")
         consulService.requestServiceStatus(id, jobServiceIds)
-      case Failure(throwable) =>
+        jobServiceIds
+    }
+    eventuallyRequestedServices.onFailure { case throwable =>
         Logger.error(s"Requesting services for $id failed: $throwable")
     }
-    eventuallyJobServiceIds
+    eventuallyRequestedServices
   }
 
   def startJob(job: JsValue): Try[Unit] = {

--- a/app/de/frosner/broccoli/services/NomadService.scala
+++ b/app/de/frosner/broccoli/services/NomadService.scala
@@ -161,11 +161,12 @@ class NomadService @Inject()(configuration: Configuration, consulService: Consul
       }
     }
     val eventuallyRequestedServices = eventuallyJobServiceIds.map { jobServiceIds =>
-        Logger.debug(s"${jobRequest.uri} => ${jobServiceIds.mkString(", ")}")
-        consulService.requestServiceStatus(id, jobServiceIds)
-        jobServiceIds
+      Logger.debug(s"${jobRequest.uri} => ${jobServiceIds.mkString(", ")}")
+      consulService.requestServiceStatus(id, jobServiceIds)
+      jobServiceIds
     }
-    eventuallyRequestedServices.onFailure { case throwable =>
+    eventuallyRequestedServices.onFailure {
+      case throwable =>
         Logger.error(s"Requesting services for $id failed: $throwable")
     }
     eventuallyRequestedServices

--- a/app/de/frosner/broccoli/services/NomadService.scala
+++ b/app/de/frosner/broccoli/services/NomadService.scala
@@ -121,12 +121,11 @@ class NomadService @Inject()(configuration: Configuration, consulService: Consul
       }
       case Failure(throwable: ConnectException) => {
         Logger.error(
-          s"Failed to request statuses for ${instanceIds.mkString(", ")} from ${jobsRequest.uri}: $throwable")
+          s"Nomad did not respond when requesting services for ${instanceIds.mkString(", ")} from ${jobsRequest.uri}: $throwable")
         setNomadNotReachable()
       }
       case Failure(throwable) => {
-        Logger.error(
-          s"Failed to request statuses for ${instanceIds.mkString(", ")} from ${jobsRequest.uri}: $throwable")
+        Logger.warn(s"Failed to request statuses for ${instanceIds.mkString(", ")} from ${jobsRequest.uri}: $throwable")
       }
     }
   }

--- a/app/de/frosner/broccoli/services/NomadService.scala
+++ b/app/de/frosner/broccoli/services/NomadService.scala
@@ -1,5 +1,6 @@
 package de.frosner.broccoli.services
 
+import java.net.ConnectException
 import java.util.concurrent.TimeUnit
 import javax.inject.{Inject, Named, Singleton}
 import javax.xml.ws.http.HTTPException
@@ -12,7 +13,7 @@ import play.api.libs.json.{JsArray, JsObject, JsString, JsValue}
 import play.api.libs.ws.WSClient
 import play.api.Configuration
 
-import scala.concurrent.Await
+import scala.concurrent.{Await, Future}
 import scala.concurrent.duration.Duration
 import scala.util.{Failure, Success, Try}
 
@@ -117,10 +118,14 @@ class NomadService @Inject()(configuration: Configuration, consulService: Consul
         jobStatuses = idsAndStatusesWithPeriodic.toMap
         filteredIdsAndStatuses.map { case (key, value) => key }.foreach(requestServices)
       }
-      case Failure(throwable) => {
+      case Failure(throwable: ConnectException) => {
         Logger.error(
           s"Failed to request statuses for ${instanceIds.mkString(", ")} from ${jobsRequest.uri}: $throwable")
         setNomadNotReachable()
+      }
+      case Failure(throwable) => {
+        Logger.error(
+          s"Failed to request statuses for ${instanceIds.mkString(", ")} from ${jobsRequest.uri}: $throwable")
       }
     }
   }
@@ -129,7 +134,7 @@ class NomadService @Inject()(configuration: Configuration, consulService: Consul
   // TODO something is stopped (i.e. not present in Nomad). Or we use a cache that let's the stuff expire if you don't get an answer from Nomad
   // https://www.playframework.com/documentation/2.5.x/ScalaCache
 
-  def requestServices(id: String): Unit = {
+  def requestServices(id: String): Future[Seq[String]] = {
     val queryUrl = nomadBaseUrl + s"/v1/job/$id"
     val jobRequest = ws.url(queryUrl)
     val jobResponse = jobRequest.get().map { response =>
@@ -140,8 +145,18 @@ class NomadService @Inject()(configuration: Configuration, consulService: Consul
       }
     }
     val eventuallyJobServiceIds = jobResponse.map { jsObject =>
-      val services = (jsObject \\ "Services").flatMap(_.as[JsArray].value.map(_.as[JsObject]))
-      services.flatMap { serviceJsObject =>
+      val services = (jsObject \\ "Services")
+      val serviceObjects = services.flatMap { serviceJson =>
+        val validatedServices = serviceJson.validate[JsArray].asEither
+        validatedServices match {
+          case Left(problem) =>
+            Logger.warn(s"Error when parsing service JSON in job $id: $problem")
+            List.empty
+          case Right(array) =>
+            array.value.map(_.as[JsObject])
+        }
+      }
+      serviceObjects.flatMap { serviceJsObject =>
         serviceJsObject.value.get("Name").map(_.as[JsString].value)
       }
     }
@@ -149,11 +164,10 @@ class NomadService @Inject()(configuration: Configuration, consulService: Consul
       case Success(jobServiceIds) =>
         Logger.debug(s"${jobRequest.uri} => ${jobServiceIds.mkString(", ")}")
         consulService.requestServiceStatus(id, jobServiceIds)
-        setNomadReachable()
       case Failure(throwable) =>
         Logger.error(s"Requesting services for $id failed: $throwable")
-        setNomadNotReachable()
     }
+    eventuallyJobServiceIds
   }
 
   def startJob(job: JsValue): Try[Unit] = {

--- a/elm/tests/elm-package.json
+++ b/elm/tests/elm-package.json
@@ -12,6 +12,7 @@
         "elm-community/json-extra": "2.0.0 <= v < 3.0.0",
         "mgold/elm-random-pcg": "4.0.2 <= v < 5.0.0",
         "eeue56/elm-html-test": "3.0.0 <= v <= 3.0.0",
+        "eeue56/elm-html-in-elm": "5.0.0 <= v <= 5.0.0",
         "elm-community/elm-test": "3.0.0 <= v < 4.0.0",
         "rtfeldman/node-test-runner": "3.0.0 <= v < 4.0.0",
         "elm-lang/core": "5.0.0 <= v < 6.0.0",

--- a/test/de/frosner/broccoli/services/NomadServiceSpec.scala
+++ b/test/de/frosner/broccoli/services/NomadServiceSpec.scala
@@ -4,6 +4,7 @@ import java.util.concurrent.TimeUnit
 
 import de.frosner.broccoli.controllers.ServiceMocks
 import de.frosner.broccoli.conf
+import de.frosner.broccoli.nomad.models.{Job, Service, Task, TaskGroup}
 import org.specs2.mutable.Specification
 import org.mockito.Mockito._
 import play.api.Configuration
@@ -12,6 +13,8 @@ import play.api.mvc._
 import play.api.routing.sird._
 import play.api.test._
 import play.core.server.Server
+
+import Job.jobFormat
 
 import scala.concurrent.Await
 import scala.concurrent.duration.Duration
@@ -23,31 +26,15 @@ class NomadServiceSpec extends Specification with ServiceMocks {
   "Requesting services for specific job" should {
 
     "ask consul for the services that nomad returns" in {
-      val service1Name = "my-service"
-      val service1 = JsObject(
-        Map(
-          "Name" -> JsString(service1Name)
-        ))
-      val services = JsArray(Seq(service1))
-      val task1 = JsObject(
-        Map(
-          "Services" -> services
-        ))
-      val tasks = JsArray(Seq(task1))
-      val taskGroup1 = JsObject(
-        Map(
-          "Tasks" -> tasks
-        ))
-      val taskGroups = JsArray(Seq(taskGroup1))
-      val job = JsObject(
-        Map(
-          "TaskGroups" -> taskGroups
-        ))
+      val service1 = Service("my-service")
+      val task1 = Task(Some(Seq(service1)))
+      val taskGroup1 = TaskGroup(Some(Seq(task1)))
+      val job = Job(Some(Seq(taskGroup1)))
       val consulService = mock(classOf[ConsulService])
       Server.withRouter() {
         case GET(p"/v1/job/my-job") =>
           Action {
-            Results.Ok(job)
+            Results.Ok(Json.toJson(job))
           }
       } { implicit port =>
         WsTestClient.withClient { client =>
@@ -58,48 +45,24 @@ class NomadServiceSpec extends Specification with ServiceMocks {
           val nomadService = new NomadService(configuration, consulService, client)
           val jobId = "my-job"
           val result = Await.result(nomadService.requestServices(jobId), Duration(5, TimeUnit.SECONDS))
-          verify(consulService).requestServiceStatus(jobId, Iterable(service1Name))
-          result === Seq(service1Name)
+          verify(consulService).requestServiceStatus(jobId, Iterable(service1.name))
+          result === Seq(service1.name)
         }
       }
     }
 
     "not explode when receiving tasks without services" in {
-      val service1Name = "my-service"
-      val service1 = JsObject(
-        Map(
-          "Name" -> JsString(service1Name)
-        ))
-      val services = JsArray(Seq(service1))
-      val task1 = JsObject(
-        Map(
-          "Services" -> services
-        ))
-      val tasks = JsArray(Seq(task1))
-      val taskGroup1 = JsObject(
-        Map(
-          "Tasks" -> tasks
-        ))
-      val services2 = JsNull
-      val task2 = JsObject(
-        Map(
-          "Services" -> services2
-        ))
-      val tasks2 = JsArray(Seq(task2))
-      val taskGroup2 = JsObject(
-        Map(
-          "Tasks" -> tasks2
-        ))
-      val taskGroups = JsArray(Seq(taskGroup1, taskGroup2))
-      val job = JsObject(
-        Map(
-          "TaskGroups" -> taskGroups
-        ))
+      val service1 = Service("my-service")
+      val task1 = Task(Some(Seq(service1)))
+      val task2 = Task(None)
+      val taskGroup1 = TaskGroup(Some(Seq(task1)))
+      val taskGroup2 = TaskGroup(Some(Seq(task2)))
+      val job = Job(Some(Seq(taskGroup1, taskGroup2)))
       val consulService = mock(classOf[ConsulService])
       Server.withRouter() {
         case GET(p"/v1/job/my-job") =>
           Action {
-            Results.Ok(job)
+            Results.Ok(Json.toJson(job))
           }
       } { implicit port =>
         WsTestClient.withClient { client =>
@@ -110,8 +73,8 @@ class NomadServiceSpec extends Specification with ServiceMocks {
           val nomadService = new NomadService(configuration, consulService, client)
           val jobId = "my-job"
           val result = Await.result(nomadService.requestServices(jobId), Duration(5, TimeUnit.SECONDS))
-          verify(consulService).requestServiceStatus(jobId, Iterable(service1Name))
-          result === Seq(service1Name)
+          verify(consulService).requestServiceStatus(jobId, Iterable(service1.name))
+          result === Seq(service1.name)
         }
       }
     }

--- a/test/de/frosner/broccoli/services/NomadServiceSpec.scala
+++ b/test/de/frosner/broccoli/services/NomadServiceSpec.scala
@@ -1,0 +1,107 @@
+package de.frosner.broccoli.services
+
+import java.util.concurrent.TimeUnit
+
+import de.frosner.broccoli.controllers.ServiceMocks
+import de.frosner.broccoli.conf
+import org.specs2.mutable.Specification
+import org.mockito.Mockito._
+import play.api.Configuration
+import play.api.libs.json._
+import play.api.mvc._
+import play.api.routing.sird._
+import play.api.test._
+import play.core.server.Server
+
+import scala.concurrent.Await
+import scala.concurrent.duration.Duration
+
+class NomadServiceSpec extends Specification with ServiceMocks {
+
+  sequential
+
+  "Requesting services for specific job" should {
+
+    "ask consul for the services that nomad returns" in {
+      val service1Name = "my-service"
+      val service1 = JsObject(Map(
+        "Name" -> JsString(service1Name)
+      ))
+      val services = JsArray(Seq(service1))
+      val task1 = JsObject(Map(
+        "Services" -> services
+      ))
+      val tasks = JsArray(Seq(task1))
+      val taskGroup1 = JsObject(Map(
+        "Tasks" -> tasks
+      ))
+      val taskGroups = JsArray(Seq(taskGroup1))
+      val job = JsObject(Map(
+        "TaskGroups" -> taskGroups
+      ))
+      val consulService = mock(classOf[ConsulService])
+      Server.withRouter() {
+        case GET(p"/v1/job/my-job") => Action {
+          Results.Ok(job)
+        }
+      } { implicit port =>
+        WsTestClient.withClient { client =>
+          val configuration = Configuration.from(Map(
+            conf.NOMAD_URL_KEY -> s"http://localhost:$port"
+          ))
+          val nomadService = new NomadService(configuration, consulService, client)
+          val jobId = "my-job"
+          val result = Await.result(nomadService.requestServices(jobId), Duration(5, TimeUnit.SECONDS))
+          verify(consulService).requestServiceStatus(jobId, Iterable(service1Name))
+          result === Seq(service1Name)
+        }
+      }
+    }
+
+    "not explode when receiving tasks without services" in {
+      val service1Name = "my-service"
+      val service1 = JsObject(Map(
+        "Name" -> JsString(service1Name)
+      ))
+      val services = JsArray(Seq(service1))
+      val task1 = JsObject(Map(
+        "Services" -> services
+      ))
+      val tasks = JsArray(Seq(task1))
+      val taskGroup1 = JsObject(Map(
+        "Tasks" -> tasks
+      ))
+      val services2 = JsNull
+      val task2 = JsObject(Map(
+        "Services" -> services2
+      ))
+      val tasks2 = JsArray(Seq(task2))
+      val taskGroup2 = JsObject(Map(
+        "Tasks" -> tasks2
+      ))
+      val taskGroups = JsArray(Seq(taskGroup1, taskGroup2))
+      val job = JsObject(Map(
+        "TaskGroups" -> taskGroups
+      ))
+      val consulService = mock(classOf[ConsulService])
+      Server.withRouter() {
+        case GET(p"/v1/job/my-job") => Action {
+          Results.Ok(job)
+        }
+      } { implicit port =>
+        WsTestClient.withClient { client =>
+          val configuration = Configuration.from(Map(
+            conf.NOMAD_URL_KEY -> s"http://localhost:$port"
+          ))
+          val nomadService = new NomadService(configuration, consulService, client)
+          val jobId = "my-job"
+          val result = Await.result(nomadService.requestServices(jobId), Duration(5, TimeUnit.SECONDS))
+          verify(consulService).requestServiceStatus(jobId, Iterable(service1Name))
+          result === Seq(service1Name)
+        }
+      }
+    }
+
+  }
+
+}

--- a/test/de/frosner/broccoli/services/NomadServiceSpec.scala
+++ b/test/de/frosner/broccoli/services/NomadServiceSpec.scala
@@ -24,31 +24,37 @@ class NomadServiceSpec extends Specification with ServiceMocks {
 
     "ask consul for the services that nomad returns" in {
       val service1Name = "my-service"
-      val service1 = JsObject(Map(
-        "Name" -> JsString(service1Name)
-      ))
+      val service1 = JsObject(
+        Map(
+          "Name" -> JsString(service1Name)
+        ))
       val services = JsArray(Seq(service1))
-      val task1 = JsObject(Map(
-        "Services" -> services
-      ))
+      val task1 = JsObject(
+        Map(
+          "Services" -> services
+        ))
       val tasks = JsArray(Seq(task1))
-      val taskGroup1 = JsObject(Map(
-        "Tasks" -> tasks
-      ))
+      val taskGroup1 = JsObject(
+        Map(
+          "Tasks" -> tasks
+        ))
       val taskGroups = JsArray(Seq(taskGroup1))
-      val job = JsObject(Map(
-        "TaskGroups" -> taskGroups
-      ))
+      val job = JsObject(
+        Map(
+          "TaskGroups" -> taskGroups
+        ))
       val consulService = mock(classOf[ConsulService])
       Server.withRouter() {
-        case GET(p"/v1/job/my-job") => Action {
-          Results.Ok(job)
-        }
+        case GET(p"/v1/job/my-job") =>
+          Action {
+            Results.Ok(job)
+          }
       } { implicit port =>
         WsTestClient.withClient { client =>
-          val configuration = Configuration.from(Map(
-            conf.NOMAD_URL_KEY -> s"http://localhost:$port"
-          ))
+          val configuration = Configuration.from(
+            Map(
+              conf.NOMAD_URL_KEY -> s"http://localhost:$port"
+            ))
           val nomadService = new NomadService(configuration, consulService, client)
           val jobId = "my-job"
           val result = Await.result(nomadService.requestServices(jobId), Duration(5, TimeUnit.SECONDS))
@@ -60,39 +66,47 @@ class NomadServiceSpec extends Specification with ServiceMocks {
 
     "not explode when receiving tasks without services" in {
       val service1Name = "my-service"
-      val service1 = JsObject(Map(
-        "Name" -> JsString(service1Name)
-      ))
+      val service1 = JsObject(
+        Map(
+          "Name" -> JsString(service1Name)
+        ))
       val services = JsArray(Seq(service1))
-      val task1 = JsObject(Map(
-        "Services" -> services
-      ))
+      val task1 = JsObject(
+        Map(
+          "Services" -> services
+        ))
       val tasks = JsArray(Seq(task1))
-      val taskGroup1 = JsObject(Map(
-        "Tasks" -> tasks
-      ))
+      val taskGroup1 = JsObject(
+        Map(
+          "Tasks" -> tasks
+        ))
       val services2 = JsNull
-      val task2 = JsObject(Map(
-        "Services" -> services2
-      ))
+      val task2 = JsObject(
+        Map(
+          "Services" -> services2
+        ))
       val tasks2 = JsArray(Seq(task2))
-      val taskGroup2 = JsObject(Map(
-        "Tasks" -> tasks2
-      ))
+      val taskGroup2 = JsObject(
+        Map(
+          "Tasks" -> tasks2
+        ))
       val taskGroups = JsArray(Seq(taskGroup1, taskGroup2))
-      val job = JsObject(Map(
-        "TaskGroups" -> taskGroups
-      ))
+      val job = JsObject(
+        Map(
+          "TaskGroups" -> taskGroups
+        ))
       val consulService = mock(classOf[ConsulService])
       Server.withRouter() {
-        case GET(p"/v1/job/my-job") => Action {
-          Results.Ok(job)
-        }
+        case GET(p"/v1/job/my-job") =>
+          Action {
+            Results.Ok(job)
+          }
       } { implicit port =>
         WsTestClient.withClient { client =>
-          val configuration = Configuration.from(Map(
-            conf.NOMAD_URL_KEY -> s"http://localhost:$port"
-          ))
+          val configuration = Configuration.from(
+            Map(
+              conf.NOMAD_URL_KEY -> s"http://localhost:$port"
+            ))
           val nomadService = new NomadService(configuration, consulService, client)
           val jobId = "my-job"
           val result = Await.result(nomadService.requestServices(jobId), Duration(5, TimeUnit.SECONDS))

